subsys/lorawan/nvm/lorawan_nvm_settings.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2021 Intellinium <giuliano.franchetto@intellinium.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <LoRaMac.h>
+#include <settings/settings.h>
+#include "lorawan_nvm.h"
+#include <logging/log.h>
+#if defined(CONFIG_LORAWAN_SE_HAS_OWN_NVM)
+#include "../se/lorawan_se.h"
+#endif
+
+LOG_MODULE_REGISTER(lorawan_nvm, CONFIG_LORAWAN_LOG_LEVEL);
+
+static uint16_t nvm_notify_flag;
+
+#define LORAWAN_SETTINGS_BASE                "lorawan/nvm"
+
+#define flag_set(f)        ((nvm_notify_flag & (f)) == (f))
+
+#define SAVE_SETTINGS(nvm, flag, name)					\
+if (flag_set(flag)) {							\
+	LOG_DBG("Saving configuration " LORAWAN_SETTINGS_BASE "/" # name);\
+	int err = settings_save_one(LORAWAN_SETTINGS_BASE "/" # name,	\
+				&(nvm)->name,				\
+				sizeof((nvm)->name));			\
+	if (err) {							\
+		LOG_ERR("Could not save settings" # name", error %d",	\
+			err);						\
+	}								\
+}
+
+static void nvm_settings_work(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	MibRequestConfirm_t mibReq;
+
+	LOG_DBG("Saving LoRaWAN settings");
+
+	/* Retrieve the actual context */
+	mibReq.Type = MIB_NVM_CTXS;
+	if (LoRaMacMibGetRequestConfirm(&mibReq) != LORAMAC_STATUS_OK) {
+		LOG_ERR("Could not get NVM context");
+		return;
+	}
+
+	LoRaMacNvmData_t *nvm = mibReq.Param.Contexts;
+
+	LOG_DBG("Crypto version: %"PRIu32", DevNonce: %d, JoinNonce: %"PRIu32,
+		mibReq.Param.Contexts->Crypto.LrWanVersion.Value,
+		mibReq.Param.Contexts->Crypto.DevNonce,
+		mibReq.Param.Contexts->Crypto.JoinNonce);
+
+	/* Wait for LoRa to be idle */
+	while (LoRaMacStop() != LORAMAC_STATUS_OK) {
+		k_sleep(K_MSEC(100));
+	}
+
+	SAVE_SETTINGS(nvm, LORAMAC_NVM_NOTIFY_FLAG_CRYPTO, Crypto)
```

Comment:
Don't we need locking here as restore can happen in parallel?

Reply:
```lorawan_nvm_data_restore()``` is only called when ```lorawan_init()``` is called.
As for ```lorawan_nvm_data_mgmt_event()```, I think the race condition is prevented thanks to [that](https://github.com/zephyrproject-rtos/zephyr/pull/40193#discussion_r785417057).